include/lbann/layers/transform/hadamard.hpp

```diff
@@ -96,159 +74,46 @@ class hadamard_layer : public transform_layer {
 
   protected:
 
-  void setup_pointers() override {
-    transform_layer::setup_pointers();
-    std::stringstream err;
-    if (get_num_parents() <= 0) {
-      err << __FILE__ << " " << __LINE__ << " :: hadamard_layer: "
-          << "Hadamard layer has no parents";
-      throw lbann_exception(err.str());
-    }
-  }
-
   void fp_compute() override {
-    if (this->m_using_gpus) {
-      fp_compute_gpu();
-    } else { 
-      fp_compute_cpu();
+    auto& output = get_activations();
+    switch (get_num_parents()) {
+    case 0: El::Fill(output, DataType(1)); break;
+    case 1: El::LockedView(output, get_prev_activations()); break;
+    default:
+      El::Hadamard(get_prev_activations(0),
+                   get_prev_activations(1),
+                   output);
+      for (int i = 2; i < get_num_parents(); ++i) {
+        El::Hadamard(get_prev_activations(i), output, output);
+      }
     }
   }
 
   void bp_compute() override {
-    if (this->m_using_gpus) {
-      bp_compute_gpu();
-    } else { 
-      bp_compute_cpu();
-    }
-  }
-
- private:
-
-  void fp_compute_cpu() {
-
-    // Return quickly if layer has fewer than two parents
-    const int num_parents = get_num_parents();
-    if (num_parents == 0) {
-      El::Fill(get_activations(), DataType(1));
-      return;
-    }
-    if (num_parents == 1) {
-      El::Copy(get_prev_activations(), get_activations());
-      return;
-    }
-
-    // Get local matrices
-    auto& local_input0 = get_local_prev_activations(0);
-    std::vector<const Mat*> local_inputs;
-    for (int i = 1; i < num_parents; ++i) {
-      local_inputs.push_back(&get_local_prev_activations(i));
-    }
-    auto& local_output = get_local_activations();
-
-    // Compute entrywise product
-    const int local_height = local_output.Height();
-    const int local_width = local_output.Width();
-    #pragma omp parallel for collapse(2)
```

Comment:
Is it safe to assume that the Elemental Hadamard is also parallelized with OMP for the CPU?